pkg/builder/entity_dto_builder.go

```diff
@@ -345,17 +401,127 @@ func (eb *EntityDTOBuilder) VirtualApplicationData(vAppData *proto.EntityDTO_Vir
 	return eb
 }
 
-func buildCommodityBoughtFromMap(providerCommoditiesMap map[string][]*proto.CommodityDTO) []*proto.EntityDTO_CommodityBought {
+// Build the commodity DTOs for the commodity bought from different providers
+func (eb *EntityDTOBuilder) buildCommodityBoughtFromMap() []*proto.EntityDTO_CommodityBought {
 	var commoditiesBought []*proto.EntityDTO_CommodityBought
-	if len(providerCommoditiesMap) == 0 {
+
+	if len(eb.commoditiesBoughtProviderMap) == 0 {
 		return commoditiesBought
 	}
-	for providerId, commodities := range providerCommoditiesMap {
+
+	var actionEligibilityMap map[proto.EntityDTO_EntityType]*ActionEligibilityByProvider
+	if eb.actionEligibility != nil {
+		actionEligibilityMap = eb.actionEligibility.actionEligibilityByProviderMap
+	}
+
+	for providerId, commodities := range eb.commoditiesBoughtProviderMap {
+		// Commodity bought per provider
 		p := providerId
-		commoditiesBought = append(commoditiesBought, &proto.EntityDTO_CommodityBought{
+
+		commodityBought := &proto.EntityDTO_CommodityBought{
 			ProviderId: &p,
 			Bought:     commodities,
-		})
+		}
+
+		// update provider type and the action eligibility of the entity across that provider type
+		providerType, providerTypeExists := eb.providerMap[p]
+		if providerTypeExists {
+
+			commodityBought.ProviderType = &providerType
+			aeMap, exists := actionEligibilityMap[providerType]
```

Comment:
nit: I don't think "aeMap" is actually a map. It looks like it's a pointer to an ActionEligibilityByProvider struct which makes this code a little bit confusing if you think it's a map.

Reply:
Incorporated in new updated PR#81